src/lib/reasoners/relation.ml

```diff
@@ -131,14 +133,31 @@ let case_split env uf ~for_model =
   let seq5 = Rel5.case_split env.r5 uf ~for_model in
   let seq6 = Rel6.case_split env.r6 uf ~for_model in
   let seq7 = Rel7.case_split env.r7 uf ~for_model in
-  let l = seq1 |@| seq2 |@| seq3 |@| seq4 |@| seq5 |@| seq6 |@| seq7 in
-  List.sort
-    (fun (_,_,sz1) (_,_,sz2) ->
+  let splits = [seq1; seq2; seq3; seq4; seq5; seq6; seq7] in
+  let splits = List.fold_left (|@|) [] splits in
+  List.fast_sort
+    (fun (_ ,_ , sz1) (_ ,_ , sz2) ->
        match sz1, sz2 with
-       | Th_util.CS(_,sz1), Th_util.CS(_,sz2) -> Numbers.Q.compare sz1 sz2
+       | Th_util.CS (_ , sz1), Th_util.CS (_ , sz2) ->
+         Numbers.Q.compare sz1 sz2
        | _ -> assert false
-    )l
+    ) splits
 
+let optimizing_split env uf ({ Th_util.r; _ } as opt_split) =
+  Options.exec_thread_yield ();
+  match X.type_info r with
+  | Ty.Tint | Ty.Treal ->
+    Rel1.optimizing_split env.r1 uf opt_split
+  | Ty.Trecord _ ->
+    Rel2.optimizing_split env.r2 uf opt_split
+  | Ty.Tbitv _ ->
+    Rel3.optimizing_split env.r3 uf opt_split
+  | Ty.Tfarray _ ->
+    Rel4.optimizing_split env.r4 uf opt_split
+  | Ty.Tadt _ ->
+    Rel5.optimizing_split env.r5 uf opt_split
+  | _ ->
+    opt_split
```

Comment:
In the previous code, we run `RelX.optimizing_split` on each split and stop after getting a `Some _` value. I changed it because in most of the codebase, we decide which of theories will manipulate a formula according to the type of its top symbol.